extension/partiql-extension-ion/src/boxed_ion.rs

```diff
@@ -531,6 +545,142 @@ enum BoxedIonValue {
     Sequence(Sequence),
 }
 
+impl Hash for BoxedIonValue {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        match self {
+            BoxedIonValue::Stream() => {
+                todo!("stream not hashable? ")
+            }
+            BoxedIonValue::Value(val) => {
+                let sha = ion_rs::ion_hash::sha256(val).expect("ion hash");
+                state.write(&sha);
+            }
+            BoxedIonValue::Sequence(_) => todo!("ion seq hash"),
+        }
+    }
+}
+
+impl<'a, const NULLS_EQUAL: bool, const NAN_EQUAL: bool> NullableEq
+    for IonEqualityValue<'a, NULLS_EQUAL, NAN_EQUAL, BoxedIonValue>
+{
+    #[inline(always)]
+    fn eq(&self, other: &Self) -> Value {
+        let wrap = IonEqualityValue::<'a, { NULLS_EQUAL }, { NAN_EQUAL }, Element>;
+        let wrap_seq = IonEqualityValue::<'a, { NULLS_EQUAL }, { NAN_EQUAL }, Sequence>;
+        match (self.0, other.0) {
+            (BoxedIonValue::Value(l), BoxedIonValue::Value(r)) => {
+                NullableEq::eq(&wrap(l), &wrap(r))
+            }
+            (BoxedIonValue::Sequence(l), BoxedIonValue::Sequence(r)) => {
+                NullableEq::eq(&wrap_seq(l), &wrap_seq(r))
+            }
+            _ => Value::Boolean(false),
+        }
+    }
+
+    #[inline(always)]
+    fn eqg(&self, rhs: &Self) -> Value {
+        let wrap = IonEqualityValue::<'_, true, { NAN_EQUAL }, _>;
+        NullableEq::eq(&wrap(self.0), &wrap(rhs.0))
+    }
+}
+
+impl<'a, const NULLS_EQUAL: bool, const NAN_EQUAL: bool> NullableEq
+    for IonEqualityValue<'a, NULLS_EQUAL, NAN_EQUAL, Element>
+{
+    fn eq(&self, other: &Self) -> Value {
+        let wrap_seq = IonEqualityValue::<'a, { NULLS_EQUAL }, { NAN_EQUAL }, Sequence>;
+        let wrap_struct = IonEqualityValue::<'a, { NULLS_EQUAL }, { NAN_EQUAL }, Struct>;
+        let (l, r) = (self.0, other.0);
+        let (lty, rty) = (l.ion_type(), r.ion_type());
+
+        let result = if l.is_null() && r.is_null() {
+            NULLS_EQUAL
+        } else {
+            match (lty, rty) {
+                (IonType::Float, IonType::Float) => {
+                    let (l, r) = (l.as_float().unwrap(), r.as_float().unwrap());
+                    if l.is_nan() && r.is_nan() {
+                        NAN_EQUAL
```

Comment:
Do annotation need to be considered for `NULL` and `NAN` equality?

Reply:
Likely so; I'll fix